Adds 16bpp support for BMP File Format

ScorpionInc opened the pull request:
This commit adds some basic 16bpp support for BMP File Format.

Doesn't support R5G6B5 pixel format yet as I haven't figured out how to detect it's usage seperatly from the A1R5G5B5/X1R5G5B5 pixel formats from the available header information. Should be easy to add support for it though once that's nailed down.

Doesn't support Alpha bit yet. This is because while it seems to still have it's transparency correct when I open test images in Gimp, I have no such luck in Godot and I'm not sure why. I think some relevant information maybe being stored in the colorspace or the V5 BMP header that I'm either ignorant of or have missed.

Edit: Added spacing and line breaks for legability.
<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->

ScorpionInc commented:
I also added the test project I was using while testing the code here: https://github.com/ScorpionInc/godot-projects/tree/main/BMPTest if that would help anyone reviewing out. It's not the prettiest project but it uses alot of different BMP image formats and colors so you can see the edge cases better.

ScorpionInc commented:
The check: "Static Checks / Static Checks (clang-format, black format, file format, documentation checks)" Fails still however the suggested results look the same as my current values. Is there some invisible format character in there somewhere that I'm missing? Thanks for your time.

ScorpionInc commented:
Alrighty should be ready now hopefully thanks as always for everyone's time and help. o/

Anutrix commented:
You can check out with my MR https://github.com/godotengine/godot/pull/67802 too.
Check if your project/files work with that MR when you can.
Based on `BI_BITFIELDS`, it should support RGB565, RGB555 or any other RGB mask. Defaults to 555 if no BI_BITFIELDS.

fire commented:
If I remember correctly, Reduz didn't like color palletizing or using formats typically LDR for HDR because it caused confusion and thus made Godot Engine less usable.

This appears to be converting RGB565, and RGB555 to RGBA8. So I think it's fine.

ScorpionInc commented:
I am not seeing where the color scaling is done. This is probably just something I missed because I'm dumb.
   I've tried compiling the code from your fork and testing it with many different 16 bit formats of BMP and have not gotten resuts I expected. This may be because I wrote my code and project for an older version of Godot and havent forwarded my code yet. However your code seems to produce the same results as stable branch but with more warnings and a black box vs more errors and transparent sprites.
   I'll try applying my code changes to the current version of Godot (4.1 at this time) and see if I get simular results, but wanted to share what I'm seeing so far.

Project:
https://github.com/ScorpionInc/godot-projects/tree/main/BMPTest
Stable Godot:
![Current_stable](https://user-images.githubusercontent.com/19487319/229607634-c79d3837-ce15-45c4-a28a-9043e7b5c426.png)
My older build:
![My_Build](https://user-images.githubusercontent.com/19487319/229607714-436b61c0-fcc5-47c7-89ba-901eed44b3fd.png)
Anutrix build:
![Anutrix_Compiled_Pull_00](https://user-images.githubusercontent.com/19487319/229607813-9201aff3-fceb-4982-9e28-702f8bc852d8.png)

   I'll keep you posted on what I find. I like the look up for masks, based on what I read in your merge it sounds like the missing piece to get some of the alpha formats working. I'm not tring to say anything bad about your code, just wanted to show I'm not seeing it working on my machine yet.
   I'll keep you guys posted on if updating the base code breaks my version as well. Have a great day o/.

ScorpionInc commented:
So applying my code to the most recent Godot build seems to break mine as well see attached.
![My_Updated_Build](https://user-images.githubusercontent.com/19487319/229632138-8c0eecf4-4a4d-4f64-a2aa-9e30573e04fc.png)
So I think there is something else going on here. Is anyone else getting these black textures for some reason? Perhaps a change in how Ref<Image>->set_data() works?

ScorpionInc commented:
So I figured out the black boxes. Because I didnt reimport and used the .import cached values instead the project was displaying the black boxes I was getting when using your build. After reimporting Mine is back to what it was before.
![Untitled3](https://user-images.githubusercontent.com/19487319/229641677-28a5d9ae-56c9-4142-b038-65f246c3c2ee.png)
However reimporting didnt change the output of your build from block boxes for some reason. I'll keep pokeing around if you'd like.

ScorpionInc committed:
Adds some basic 16BPP BMP Support.

This commit adds some basic 16bpp support for BMP File Format.

Added support for reading and using of 16 bit mask values from file. All values are scaled to ARGB255 format based on bit depth of source color channel.

Removed warning, it's no longer required as 16 bit with alpha bit(s) are now supported.

Adjusted spacing, added spacing and brackets to make clang static check happy.

ScorpionInc commented:
Thought I'd share a little update. So far I got almost all versions of 16 bit bmps working im only having issues with the X8R8G8B8 format. for some reason the red mask reads a 0x0000 instead of the expected: 0x0F00. Interestingly A8R8G8B8 also does this however it displays correctly where X8 does not. Still working on it but I have masks implements mostly and thought I'd share.

ScorpionInc commented:
![Untitled6_MyCode](https://user-images.githubusercontent.com/19487319/230018934-8275ff12-e739-41f6-a7c8-ebc79b647bc1.png)
All 16-bit formats are working. The bottom left image is a 32-bit image. Now Ready for review hopefully.
Thank you to Anutrix for his/her help getting the bitmasks.

clayjohn commented on modules/bmp/image_loader_bmp.cpp:
```diff
@@ -32,6 +32,15 @@
+	// Returns number of ones in the binary value of the parameter: mask.
+	// Uses a Simple pop_count.
+	uint8_t c = 0u;
+	for (; mask != 0u; mask &= mask - 1u)
+		c++;
```
```suggestion
	for (; mask != 0u; mask &= mask - 1u) {
		c++;
	}
```

Our style is to always use braces around ``if`` statements

ScorpionInc replied:
I added braces. Thank you for your suggestion.

clayjohn reviewed (commented):
Looks good to me. I just have one suggested change to style/formatted!

clayjohn reviewed (approved):
The changes seem fine to me. Should be good to merge this pending CI passing. Great work!

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:
